File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        if(nums.length == 1 )return 1;
        int[] dp = new int[nums.length];
        int answer = Integer.MIN_VALUE;
        for(int i = 1; i < nums.length; i++){
            for(int j = 0; j < i; j++){
                if(nums[j] < nums[i]){
                    dp[i] = Math.max(dp[i],dp[j]+1);
                }
            }
            answer = Math.max(answer,dp[i]);
        }
        return answer+1;
    }
}

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> result;
        for(int i = 0; i < nums.size(); i++){
            if(result.empty() or result.back() < nums[i]){
            result.push_back(nums[i]);
            }else{
                auto it = lower_bound(result.begin(),result.end(),nums[i]);
                *it = nums[i];
            }
        }
        return result.size();
    }
};